src/Core/Indexes/KDTree.cs

```diff
@@ -0,0 +1,724 @@
+
+            if (treeDimensions != 2 && treeDimensions != 3)
+                throw new ArgumentOutOfRangeException("treeDimensions", "The tree's dimensions must be 2 or 3.");
+
+            this.treeDimensions = treeDimensions;
+
+            this.InitializeTree(points.ToList());
+        }
+
+        /// <summary>
+        /// Initializes the tree.
+        /// </summary>
+        /// <param name="coordinates">The list of coordinates to create the tree from.</param>
+        private void InitializeTree(List<Coordinate> coordinates)
+        {
+            // If there are no coordinates specified, leave the tree empty.
+            if (coordinates.Count == 0)
+                return;
+
+            // Sorting the input points by the X coordinate
+            coordinates.Sort((p1, p2) => p1.X.CompareTo(p2.X));
+
+            // Setting up the root to be the median by the X coordinate
+            Coordinate median = coordinates[coordinates.Count / 2];
+            this.root = new KDTreeNode(median, 1, this.TreeDimension);
+
+            // Add all input points to the tree
+            this.root.InitializeSubTree(coordinates.GetRange(0, coordinates.Count / 2), coordinates.GetRange(coordinates.Count / 2 + 1, (int)Math.Ceiling(coordinates.Count / 2.0) - 1));
+        }
+
+        /// <summary>
+        /// Gets the tree's dimension.
+        /// </summary>
+        public int TreeDimension { get { return this.treeDimensions; } }
+
+        /// <summary>
+        /// Gets a value indicating whether the tree is empty.
+        /// </summary>
+        public bool IsEmpty { get { return this.root == null; } }
+
+        /// <summary>
+        /// Gets a value indicating whether the tree is readonly.
+        /// </summary>
+        public bool IsReadOnly
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Gets the number of geometries stored in the tree.
+        /// </summary>
+        public int NumberOfGeometries
+        {
+            get
+            {
+                return this.IsEmpty ? 0 : this.root.NumberOfGeometries;
+            }
+        }
+
+        /// <summary>
+        /// Adds a coordinate to the tree.
+        /// </summary>
+        /// <param name="coordinate">The coordinate to be added.</param>
+        /// <exception cref="System.ArgumentNullException">The coordinate is null.</exception>
+        /// <exception cref="System.InvalidOperationException">The tree is read-only.</exception>
+        /// <exception cref="System.ArgumentException">The coordinate is already in the tree.</exception>
+        public void Add(Coordinate coordinate)
+        {
+            if (coordinate == null)
+                throw new ArgumentNullException("coordinate", "The coordinate is null.");
+            if (this.IsReadOnly)
+                throw new InvalidOperationException("The tree is read-only. Adding is not allowed.");
```

Comment:
I thought I would leave this in just in case IsReadOnly will get changed in the future or in case of some inheritance where the property will not be a constant. That being said, this probably shouldn't be here.